api/include/opentelemetry/trace/propagation/HttpTraceContext.cc

```diff
@@ -0,0 +1,129 @@
+                TraceId traceId = TraceId.fromLowerBase16(traceparent, TRACE_ID_OFFSET);
+                SpanId spanId = SpanId.fromLowerBase16(traceparent, SPAN_ID_OFFSET);
+                TraceFlags traceFlags = TraceFlags.fromLowerBase16(traceparent, TRACE_OPTION_OFFSET);
+                return SpanContext.createFromRemoteParent(traceId, spanId, traceFlags, TRACE_STATE_DEFAULT);
+            } catch (IllegalArgumentException e) {
+                std::cout<<"Unparseable traceparent header. Returning INVALID span context."<<std::endl;
+                return SpanContext.getInvalid();
+            }
+        }
+
+        static TraceState extractTraceState(nostd::string_view traceStateHeader) {
+            // TODO: implementation
+            /** The question here is that how should I treat the trace state. Implementation will differ pretty much
+                depending on what is the data structure of TraceState
+            */
+        }
+
+        static SpanContext extractImpl(Getter getter, T &carrier) {
+            nostd::string_view traceParent = getter(carrier, TRACE_PARENT);
+            if (traceParent == NULL) {
+              return SpanContext.getInvalid(); // I need support of this method from SpanContext definition which is currently unavailable it seems
+            }
+
+            SpanContext contextFromParentHeader = extractContextFromTraceParent(traceParent);
+            if (!contextFromParentHeader.isValid()) {
+                // Again, I still need the support of this method from context.
+                return contextFromParentHeader;
+            }
+
+            nostd::string_view traceStateHeader = getter(carrier, TRACE_STATE);
+            if (traceStateHeader == NULL || traceStateHeader.isEmpty()) {
+                return contextFromParentHeader;
+            }
+
+            try {
+                // Again, the trace state problem, should I treat it like a string or dictionary or an encapsuled class? This is not defined in current files
+                TraceState traceState = extractTraceState(traceStateHeader);
+                // Need getter support from SpanContext
+                return SpanContext.createFromRemoteParent(
+                    contextFromParentHeader.getTraceId(),
+                    contextFromParentHeader.getSpanId(),
+                    contextFromParentHeader.getTraceFlags(),
+                    traceState);
+            } catch (IllegalArgumentException e) {
+              std::cout<<"Unparseable tracestate header. Returning span context without state."<<std::endl;
+              return contextFromParentHeader;
+            }
+        }
+    public:
+        List<String> fields() {
+            return FIELDS;
+        }
+
+        void inject(Setter setter, T &carrier, const Context &context) {
+            checkNotNull(context, "context");
+            checkNotNull(setter, "setter");
```

Comment:
references cannot be null.

(Also, checkNotNull is defined to take T, which a single type and wouldn't likely apply to both T and Context).